File: app/domain/stringify_ast.py
"""
Canonical AST ↔ string conversion for Notees (Python backend).

Two symmetrical entry points:

    ``stringify_ast(ast, options)``  – AST → readable string
    ``parse_ast(value, mode)``      – input → AST objects

``stringify_ast`` modes (closed enum — ``StringifyMode``):
    NODE_MARKDOWN   – preserves node semantics ([[…]], [label]([[…]]))
    PLAIN_MARKDOWN  – standard portable Markdown, no [[…]]
    TEXT_ONLY        – plain text for search / indexing

``parse_ast`` modes (closed enum — ``ParseMode``):
    JSON       – JSON string or list → validated AST (default)
    PLAIN      – wrap plain text as-is in a paragraph
    MARKDOWN   – parse inline Markdown (bold, italic, code, …)

Helper:
    ``serialize_ast(ast)``  – ``json.dumps`` shortcut for DB storage

Node links are resolved recursively with cycle detection.

Notees
Copyright (C) 2026 Miquel Rosell Tarragó
AGPL-3.0 – see LICENSE.
"""
from __future__ import annotations

import json
import re
from enum import Enum
from dataclasses import dataclass
from typing import (
    Any,
    Callable,
    List,
    Optional,
    Sequence,
    Set,
)

__all__ = [
    "StringifyMode",
    "ParseMode",
    "NodeLinkResolution",
    "NodeLinkResolver",
    "StringifyOptions",
    "stringify_ast",
    "parse_ast",
    "serialize_ast",
]


# ── Public types ────────────────────────────────────────────────────


class StringifyMode(str, Enum):
    """Closed enum of stringify modes (AST → string)."""

    NODE_MARKDOWN = "NODE_MARKDOWN"
    """Internal Markdown that preserves node semantics."""

    PLAIN_MARKDOWN = "PLAIN_MARKDOWN"
    """Standard Markdown without node semantics (for export)."""

    TEXT_ONLY = "TEXT_ONLY"
    """Plain text for search indexing."""


class ParseMode(str, Enum):
    """Closed enum of parse modes (string → AST)."""

    JSON = "JSON"
    """Deserialize a JSON string (or pass through a list). Default."""

    PLAIN = "PLAIN"
    """Wrap plain text as-is in a single text node (no formatting)."""

    MARKDOWN = "MARKDOWN"
    """Parse inline Markdown: **bold**, *italic*, `code`, ~~strike~~, ==highlight==, [text](url)."""


@dataclass(frozen=True, slots=True)
class NodeLinkResolution:
    """Result of resolving a node link."""

    target_ast: List[dict]
    """The target node's name AST."""

    label: Optional[str]
    """Fallback label (may be None; AST label takes precedence)."""

    target_id: str
    """Opaque node identifier for cycle detection."""

    is_page: Optional[bool] = None
    """Whether the target node is a page (True) or a block (False/None)."""


# Callable that takes a link_id (node_link UUID) and returns resolution or None.
NodeLinkResolver = Callable[[str], Optional[NodeLinkResolution]]


@dataclass(frozen=True, slots=True)
class StringifyOptions:
    """Options for stringify_ast."""

    mode: StringifyMode
    max_length: Optional[int] = None
    resolve_node_link: Optional[NodeLinkResolver] = None
    # When True (HTML export), node_link nodes emit [text](#target-uuid) so
    # _markdown_inline_to_html converts them to clickable <a> elements.
    html_anchors: bool = False

    # Internal — callers should NOT set this.
    _visited: frozenset[str] = frozenset()


# ── Entry point ─────────────────────────────────────────────────────


def stringify_ast(
    ast: Sequence[dict],
    options: StringifyOptions,
) -> str:
    """Stringify a complete AST document.

    Deterministic.  Side-effect-free.  No global state.
    """
    result = _render_document(list(ast), options)
    if options.max_length is not None and len(result) > options.max_length:
        return result[: options.max_length]
    return result


def parse_ast(value: Any, mode: ParseMode = ParseMode.JSON) -> List[dict]:
    """Parse input into a validated AST document.

    Modes:
        JSON (default):
            - JSON string → parse → validate
            - list         → validate
            - anything else → empty document
        PLAIN:
            - Wrap text as-is in ``[{paragraph: [{text: value}]}]``
            - Empty / None → empty document
        MARKDOWN:
            - Parse inline Markdown formatting into AST nodes
            - Supports: **bold**, *italic*, `code`, ~~strike~~,
              ==highlight==, [text](url)
            - Empty / None → empty document
    """
    if mode is ParseMode.JSON:
        return _parse_json(value)

    # PLAIN and MARKDOWN both need a string
    if not isinstance(value, str) or not value:
        return []

    if mode is ParseMode.PLAIN:
        return [{"type": "paragraph", "children": [{"type": "text", "text": value}]}]

    # MARKDOWN
    return _parse_markdown(value)


def serialize_ast(ast: List[dict]) -> str:
    """Serialize an AST document to a JSON string for DB storage.

    This is a thin wrapper around ``json.dumps`` that pairs with
    ``parse_ast(..., ParseMode.JSON)`` for deserialization.
    """
    return json.dumps(ast)


# ── Document / block rendering ─────────────────────────────────────


def _render_document(blocks: List[dict], opts: StringifyOptions) -> str:
    if not blocks:
        return ""

    is_text = opts.mode is StringifyMode.TEXT_ONLY
    rendered = [_render_block(b, opts) for b in blocks]

    if is_text:
        return _collapse_whitespace(" ".join(rendered))

    return "\n\n".join(rendered)


def _render_block(block: dict, opts: StringifyOptions) -> str:
    block_type = block.get("type")
    if block_type == "paragraph":
        return _render_inline_sequence(block.get("children", []), opts)
    # Unknown block type — stable placeholder.
    return ""


# ── Inline rendering ───────────────────────────────────────────────


def _render_inline_sequence(nodes: List[dict], opts: StringifyOptions) -> str:
    return "".join(_render_inline(n, opts) for n in nodes)


def _render_inline(node: dict, opts: StringifyOptions) -> str:
    node_type = node.get("type")
    mode = opts.mode
    is_text = mode is StringifyMode.TEXT_ONLY

    if node_type == "text":
        return node.get("text", "")

    if node_type == "hard_break":
        return " " if is_text else "  \n"

    if node_type == "strong":
        inner = _render_inline_sequence(node.get("children", []), opts)
        return inner if is_text else f"**{inner}**"

    if node_type == "em":
        inner = _render_inline_sequence(node.get("children", []), opts)
        return inner if is_text else f"*{inner}*"

    if node_type == "code":
        t = node.get("text", "")
        return t if is_text else f"`{t}`"

    if node_type == "strikethrough":
        inner = _render_inline_sequence(node.get("children", []), opts)
        return inner if is_text else f"~~{inner}~~"

    if node_type == "highlight":
        inner = _render_inline_sequence(node.get("children", []), opts)
        return inner if is_text else f"=={inner}=="

    if node_type == "underline":
        inner = _render_inline_sequence(node.get("children", []), opts)
        return inner if is_text else f"<u>{inner}</u>"

    if node_type == "external_link":
        link_text = _render_inline_sequence(node.get("children", []), opts)
        if is_text:
            return link_text
        url = node.get("url", "")
        return f"[{link_text}]({url})"

    if node_type == "node_link":
        link_id = node.get("link_id", "")
        ref_type = node.get("ref_type", "node")
        ast_label = node.get("label") or None
        return _render_node_link(link_id, ref_type, opts, ast_label=ast_label)

    # Unknown inline node — ignore silently.
    return ""


# ── Node link rendering ────────────────────────────────────────────


def _render_node_link(link_id: str, ref_type: str, opts: StringifyOptions, *, ast_label: str | None = None) -> str:
    """Render a node_link AST node according to the current mode.

    NODE_MARKDOWN:
        Without label: ``[[resolved_text]]``
        With label:    ``[label]([[resolved_text]])``

    PLAIN_MARKDOWN / TEXT_ONLY:
        label if present, otherwise resolved node text.

    AST label takes precedence over the resolution (DB) label.
    Cycle-safe: if the target was already visited, emits "…".
    """
    resolver = opts.resolve_node_link
    placeholder_md = "[[…]]" if opts.mode is StringifyMode.NODE_MARKDOWN else "…"

    if resolver is None:
        # No resolver — use AST label if available
        if ast_label:
            if opts.mode is StringifyMode.NODE_MARKDOWN:
                if ref_type == "class":
                    return ast_label
                return f"[{ast_label}]([[{link_id}]])"
            return ast_label
        return placeholder_md

    resolution = resolver(link_id)
    if resolution is None:
        if ast_label:
            if opts.mode is StringifyMode.NODE_MARKDOWN:
                if ref_type == "class":
                    return ast_label
                return f"[{ast_label}]([[{link_id}]])"
            return ast_label
        return placeholder_md

    target_ast, label, target_id = (
        resolution.target_ast,
        ast_label if ast_label is not None else resolution.label,
        resolution.target_id,
    )

    # ── Cycle detection ──
    if target_id in opts._visited:
        return placeholder_md

    child_opts = StringifyOptions(
        mode=opts.mode,
        max_length=opts.max_length,
        resolve_node_link=opts.resolve_node_link,
        html_anchors=opts.html_anchors,
        _visited=opts._visited | {target_id},
    )

    resolved_text = stringify_ast(target_ast, child_opts)

    if opts.mode is StringifyMode.NODE_MARKDOWN:
        if ref_type == "class":
            return label if label else f"{{{{{resolved_text}}}}}"
        # Node references
        if label:
            return f"[{label}]([[{resolved_text}]])"
        return f"[[{resolved_text}]]"

    # PLAIN_MARKDOWN / TEXT_ONLY
    display = label if label else resolved_text
    if opts.html_anchors and opts.mode is StringifyMode.PLAIN_MARKDOWN:
        # Extract target node UUID from link_id (format: "targetUUID:linkUUID")
        colon = link_id.find(':')
        target_uuid = link_id[:colon] if colon > 0 else link_id
        if target_uuid:
            return f"[{display}](#{target_uuid})"
    if opts.mode is StringifyMode.PLAIN_MARKDOWN:
        # Markdown export: [name]([[uuid]]) for all node links
        colon = link_id.find(':')
        target_uuid = link_id[:colon] if colon > 0 else link_id
        return f"[{display}]([[{target_uuid}]])"
    return display


# ── parse_ast helpers ───────────────────────────────────────────────


def _parse_json(value: Any) -> List[dict]:
    """Parse a JSON string or list into a validated AST document."""
    if isinstance(value, str):
        if not value:
            return []
        try:
            parsed = json.loads(value)
            if not isinstance(parsed, list):
                return []
            return _validate_document(parsed)
        except (json.JSONDecodeError, TypeError):
            return []
    if isinstance(value, list):
        return _validate_document(value)
    return []


# ── Markdown inline patterns ───────────────────────────────────────
#
# We parse a flat string into inline AST nodes.  The grammar handles:
#     **bold**  *italic*  `code`  ~~strikethrough~~  ==highlight==
#     [link text](url)
#
# Nesting (e.g. **bold *and italic***) is supported one level deep.

_MD_INLINE_RE = re.compile(
    r"(?P<code>`[^`]+`)"                   # `code`
    r"|(?P<bold_italic>\*\*\*(?P<bi>.+?)\*\*\*)"  # ***bold italic***
    r"|(?P<bold>\*\*(?P<b>.+?)\*\*)"       # **bold**
    r"|(?P<italic>\*(?P<i>[^*]+?)\*)"       # *italic*
    r"|(?P<strike>~~(?P<s>.+?)~~)"          # ~~strike~~
    r"|(?P<highlight>==(?P<h>.+?)==)"        # ==highlight==
    r"|(?P<link>\[(?P<lt>[^\]]+)\]\((?P<lu>[^)]+)\))"  # [text](url)
)


def _parse_markdown(text: str) -> List[dict]:
    """Parse inline Markdown into an AST document (one paragraph)."""
    children = _parse_md_inline(text)
    if not children:
        return []
    return [{"type": "paragraph", "children": children}]


def _parse_md_inline(text: str) -> List[dict]:
    """Parse inline Markdown patterns into a list of AST inline nodes."""
    nodes: List[dict] = []
    pos = 0

    for m in _MD_INLINE_RE.finditer(text):
        start, end = m.start(), m.end()

        # Emit plain text before this match
        if start > pos:
            nodes.append({"type": "text", "text": text[pos:start]})

        if m.group("code"):
            raw = m.group("code")
            nodes.append({"type": "code", "text": raw[1:-1]})

        elif m.group("bold_italic"):
            inner = m.group("bi")
            nodes.append({"type": "strong", "children": [
                {"type": "em", "children": [{"type": "text", "text": inner}]}
            ]})

        elif m.group("bold"):
            inner = m.group("b")
            nodes.append({"type": "strong", "children": _parse_md_inline(inner)})

        elif m.group("italic"):
            inner = m.group("i")
            nodes.append({"type": "em", "children": _parse_md_inline(inner)})

        elif m.group("strike"):
            inner = m.group("s")
            nodes.append({"type": "strikethrough", "children": _parse_md_inline(inner)})

        elif m.group("highlight"):
            inner = m.group("h")
            nodes.append({"type": "highlight", "children": _parse_md_inline(inner)})

        elif m.group("link"):
            link_text = m.group("lt")
            url = m.group("lu")
            nodes.append({"type": "external_link", "url": url, "children": [
                {"type": "text", "text": link_text}
            ]})

        pos = end

    # Trailing text
    if pos < len(text):
        nodes.append({"type": "text", "text": text[pos:]})

    return nodes


# ── Utilities ───────────────────────────────────────────────────────

_WS_RE = re.compile(r"\s+")


def _collapse_whitespace(s: str) -> str:
    return _WS_RE.sub(" ", s).strip()


def _validate_document(doc: Any) -> List[dict]:
    """Shallow validation — each element must be a dict with ``type``."""
    if not isinstance(doc, list):
        return []
    for block in doc:
        if not isinstance(block, dict) or "type" not in block:
            return []
    return doc

File: frontend/src/lib/stringifyAST.ts
/**
 * Canonical AST → string stringifier for Notees.
 *
 * ONE function — `stringifyAST` — controlled ONLY by `StringifyMode`.
 * No other stringifier may exist.
 *
 * Modes (closed enum):
 *   NODE_MARKDOWN   – preserves node semantics ([[…]], [label]([[…]]))
 *   PLAIN_MARKDOWN  – standard portable Markdown, no [[…]]
 *   TEXT_ONLY        – plain text for search / indexing
 *
 * Node links are resolved recursively with cycle detection.
 */

import type {
  ASTDocument,
  ASTBlockNode,
  ASTInlineNode,
  ASTDateRange,
} from '@/types/ast';
import { formatDateRange } from '@/utils/dateRange';

// ─── Public types ──────────────────────────────────────────────────

/** Closed enum of stringify modes. */
export const StringifyMode = {
  /** Internal Markdown that preserves node semantics. */
  NODE_MARKDOWN: 'NODE_MARKDOWN',
  /** Standard Markdown without node semantics (for export). */
  PLAIN_MARKDOWN: 'PLAIN_MARKDOWN',
  /** Plain text for search indexing. */
  TEXT_ONLY: 'TEXT_ONLY',
} as const;

export type StringifyMode = (typeof StringifyMode)[keyof typeof StringifyMode];

/**
 * Resolver callback the caller must supply so the stringifier stays
 * side-effect-free and database-agnostic.
 *
 * Given a `link_id` (node_link UUID), return:
 *   - `targetAST`  – the target node's name AST (for recursive stringification)
 *   - `label`      – optional fallback label (may be null; AST label takes precedence)
 *   - `targetId`   – opaque node identifier used only for cycle detection
 *
 * Return `null` if the link cannot be resolved (deleted node, etc.).
 */
export interface NodeLinkResolution {
  readonly targetAST: ASTDocument;
  readonly label: string | null;
  readonly targetId: string;
}

export type NodeLinkResolver = (linkId: string) => NodeLinkResolution | null;

export interface StringifyOptions {
  /** Which rendering mode to use. */
  readonly mode: StringifyMode;
  /** Truncate output to this many characters (applied at the end). */
  readonly maxLength?: number;
  /** Resolver for node links. Required for any AST that contains node_link nodes. */
  readonly resolveNodeLink?: NodeLinkResolver;
  /**
   * Internal — callers should NOT set this.
   * Tracks visited node IDs for cycle detection during recursive resolution.
   */
  readonly _visited?: ReadonlySet<string>;
}

// ─── Entry point ───────────────────────────────────────────────────

/**
 * Stringify a complete AST document.
 *
 * Deterministic. Side-effect-free. No global state.
 */
export function stringifyAST(ast: ASTDocument, options: StringifyOptions): string {
  const result = renderDocument(ast, options);
  if (options.maxLength != null && result.length > options.maxLength) {
    return result.slice(0, options.maxLength);
  }
  return result;
}

// ─── Document / block rendering ────────────────────────────────────

function renderDocument(blocks: ASTBlockNode[], opts: StringifyOptions): string {
  if (blocks.length === 0) return '';

  const isText = opts.mode === StringifyMode.TEXT_ONLY;
  const rendered = blocks.map((b) => renderBlock(b, opts));

  if (isText) {
    // TEXT_ONLY: blocks separated by single space, collapse whitespace at the end.
    return collapseWhitespace(rendered.join(' '));
  }

  // NODE_MARKDOWN / PLAIN_MARKDOWN: double-newline between paragraphs.
  return rendered.join('\n\n');
}

function renderBlock(block: ASTBlockNode, opts: StringifyOptions): string {
  switch (block.type) {
    case 'paragraph':
      return renderInlineSequence(block.children, opts);
    case 'heading':
      // Heading level is computed at render time from block depth; here we
      // just emit the inline content. Export and markdown modes add # prefix.
      return renderInlineSequence(block.children, opts);
    case 'whiteboard':
      // Whiteboard blocks render as their title + content for text/search purposes.
      {
        const title = block.title || '';
        const elements = block.data?.elements || [];
        const textParts: string[] = [];
        
        // Sort elements by Y then X to maintain logical reading order
        const sortedElements = [...elements].sort((a, b) => {
          // Group by rows (fuzzy Y sort): if Y difference is small, treat as same row
          if (Math.abs(a.y - b.y) > 20) return a.y - b.y;
          return a.x - b.x;
        });
        
        for (const el of sortedElements) {
          // Extract text from text and shape elements
          if ((el.type === 'text' || el.type === 'shape') && 'text' in el && el.text) {
             textParts.push(el.text);
          }
        }
        
        const contentText = textParts.join(' ');
        
        if (title && contentText) {
          return `${title} (${contentText})`;
        }
        
        return title || contentText;
      }
    default:
      // Unknown block type — stable placeholder.
      return '';
  }
}

// ─── Inline rendering ──────────────────────────────────────────────

function renderInlineSequence(nodes: ASTInlineNode[], opts: StringifyOptions): string {
  return nodes.map((n) => renderInline(n, opts)).join('');
}

function renderInline(node: ASTInlineNode, opts: StringifyOptions): string {
  switch (node.type) {
    case 'text':
      return node.text;

    case 'hard_break':
      return opts.mode === StringifyMode.TEXT_ONLY ? ' ' : '  \n';

    case 'code':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return node.text;
      }
      return `\`${node.text}\``;

    case 'math':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return node.expression;
      }
      return node.displayMode ? `$$${node.expression}$$` : `$${node.expression}$`;

    case 'strong':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return renderInlineSequence(node.children, opts);
      }
      return `**${renderInlineSequence(node.children, opts)}**`;

    case 'em':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return renderInlineSequence(node.children, opts);
      }
      return `*${renderInlineSequence(node.children, opts)}*`;

    case 'strikethrough':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return renderInlineSequence(node.children, opts);
      }
      return `~~${renderInlineSequence(node.children, opts)}~~`;

    case 'highlight':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return renderInlineSequence(node.children, opts);
      }
      return `==${renderInlineSequence(node.children, opts)}==`;

    case 'underline':
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return renderInlineSequence(node.children, opts);
      }
      return `__${renderInlineSequence(node.children, opts)}__`;

    case 'external_link': {
      const linkText = renderInlineSequence(node.children, opts);
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return linkText;
      }
      return `[${linkText}](${node.url})`;
    }

    case 'node_link':
      return renderNodeLink(node.link_id, node.ref_type, opts, node.label);

    case 'broken_link': {
      const text = node.label || node.link_id.split(':')[0] || '';
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return text;
      }
      return `[${text}]([broken])`;
    }

    case 'date_range': {
      const label = node.label || formatDateRange(node as ASTDateRange);
      if (opts.mode === StringifyMode.TEXT_ONLY) {
        return label;
      }
      return `[${label}]([date-range])`;
    }

    default:
      // Unknown inline node — return empty string (stable, deterministic).
      return '';
  }
}

// ─── Node link rendering ───────────────────────────────────────────

/**
 * Render a node_link AST node according to the current mode.
 *
 * NODE_MARKDOWN:
 *   Without label: [[resolved_text]]
 *   With label:    [label]([[resolved_text]])
 *
 * PLAIN_MARKDOWN / TEXT_ONLY:
 *   label if present, otherwise resolved node text.
 *
 * Cycle-safe: if the target was already visited, emits "…".
 *
 * @param astLabel - label from the AST node itself (single source of truth)
 */
function renderNodeLink(
  linkId: string,
  refType: 'node' | 'class' | 'embed' | 'user',
  opts: StringifyOptions,
  astLabel?: string | null,
): string {
  const resolver = opts.resolveNodeLink;
  if (!resolver) {
    // No resolver available — use AST label if present, otherwise stable placeholder.
    if (astLabel) return opts.mode === StringifyMode.NODE_MARKDOWN ? `[${astLabel}]([[…]])` : astLabel;
    return opts.mode === StringifyMode.NODE_MARKDOWN ? '[[…]]' : '…';
  }

  const resolution = resolver(linkId);
  if (!resolution) {
    // Link target deleted or unresolvable — use AST label if present.
    if (astLabel) return opts.mode === StringifyMode.NODE_MARKDOWN ? `[${astLabel}]([[…]])` : astLabel;
    return opts.mode === StringifyMode.NODE_MARKDOWN ? '[[…]]' : '…';
  }

  // AST label takes precedence over DB label
  const label = astLabel ?? resolution.label;
  const { targetAST, targetId } = resolution;

  // ── Cycle detection ──
  const visited = opts._visited ?? new Set<string>();
  if (visited.has(targetId)) {
    return opts.mode === StringifyMode.NODE_MARKDOWN ? '[[…]]' : '…';
  }

  // Build child options with updated visited set.
  const nextVisited = new Set(visited);
  nextVisited.add(targetId);
  const childOpts: StringifyOptions = { ...opts, _visited: nextVisited };

  // Recursively stringify the target node's AST.
  const resolvedText = stringifyAST(targetAST, childOpts);

  switch (opts.mode) {
    case StringifyMode.NODE_MARKDOWN: {
      if (refType === 'user') {
        return `[@${label ?? resolvedText}](@user)`;
      }
      if (refType === 'class') {
        // Class references: {{resolved_text}} or label
        if (label) {
          return label;
        }
        return `{{${resolvedText}}}`;
      }
      // Node references
      if (label) {
        return `[${label}]([[${resolvedText}]])`;
      }
      return `[[${resolvedText}]]`;
    }

    case StringifyMode.PLAIN_MARKDOWN:
      return refType === 'user' ? `@${label ?? resolvedText}` : (label ?? resolvedText);

    case StringifyMode.TEXT_ONLY:
      return refType === 'user' ? `@${label ?? resolvedText}` : (label ?? resolvedText);

    default:
      return label ?? resolvedText;
  }
}

// ─── Utilities ─────────────────────────────────────────────────────

function collapseWhitespace(s: string): string {
  return s.replace(/\s+/g, ' ').trim();
}
